node/src/main/scala/com/wavesplatform/transaction/smart/script/trace/TraceStep.scala

```diff
@@ -72,26 +73,62 @@ case class InvokeScriptTrace(
     v match {
       case ScriptResultV3(ds, ts) =>
         Json.obj(
-          "data" -> ds.map(
-            item =>
-              Json.obj(
-                "key"   -> item.key,
-                "value" -> item.value.toString
-              )),
-          "transfers" -> ts.map {
-            transfer =>
-              Json.obj(
-                "address" -> transfer.recipient.bytes.toString,
-                "amount"  -> transfer.amount,
-                "assetId" -> (transfer.assetId match {
-                  case Some(id) => id.toString
-                  case None     => JsNull
-                })
-              )
+          "data"      -> ds.map(dataItemJson),
+          "transfers" -> ts.map(transferJson)
+        )
+      case ScriptResultV4(actions) =>
+        Json.obj(
+          "actions" -> actions.map {
+            case transfer: AssetTransfer => transferJson(transfer) + ("type" -> JsString("transfer"))
+            case issue: Issue            => issueJson(issue)       + ("type" -> JsString("issue"))
+            case reissue: Reissue        => reissueJson(reissue)   + ("type" -> JsString("reissue"))
+            case burn: Burn              => burnJson(burn)         + ("type" -> JsString("burn"))
+            case item: DataItem[_]       => dataItemJson(item)     + ("type" -> JsString("dataItem"))
           }
         )
-      case ScriptResultV4(actions) => ??? //todo
     }
+
+  private def transferJson(transfer: AssetTransfer) =
+    Json.obj(
+      "address" -> transfer.recipient.bytes.toString,
+      "amount"  -> transfer.amount,
+      "assetId" -> (transfer.assetId match {
+        case Some(id) => id.toString
+        case None     => JsNull
+      })
+    )
+
+  private def dataItemJson(item: DataItem[_]) =
+    Json.obj(
+      "key"   -> item.key,
+      "value" -> item.value.toString
+    )
+
+  private def issueJson(issue: Issue) =
+    Json.obj(
+      "script" -> (issue.compiledScript match {
+        case Some(script) => script.toString
+        case None         => JsNull
+      }),
+      "decimals"     -> issue.decimals,
+      "description"  -> issue.description,
+      "isReissuable" -> issue.isReissuable,
+      "quantity"     -> issue.quantity,
+      "name"         -> issue.name
+    )
+
+  private def reissueJson(reissue: Reissue) =
+    Json.obj(
+      "assetId"      -> reissue.assetId.toString,
```

Comment:
is needed to rename to "id" as field `Asset.id`? How do you think?

Reply:
I would not change to make it clearer